File: ts/src/scoring/kendall_tau.ts
export function kendallTau(
  rankingA: string[],
  rankingB: string[],
): number | null {
  const bSet = new Set(rankingB);
  const common = rankingA.filter((x) => bSet.has(x));
  const n = common.length;
  if (n < 2) return null;

  const rankB = new Map<string, number>();
  for (let i = 0; i < rankingB.length; i++) {
    rankB.set(rankingB[i], i);
  }
  const ordered = common.map((item) => rankB.get(item)!);

  let concordant = 0;
  let discordant = 0;
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (ordered[i] < ordered[j]) concordant++;
      else if (ordered[i] > ordered[j]) discordant++;
    }
  }

  const pairs = (n * (n - 1)) / 2;
  if (pairs === 0) return null;
  return (concordant - discordant) / pairs;
}

File: src/synapt_eval/scoring/kendall_tau.py
"""Kendall's Tau rank correlation coefficient."""

from __future__ import annotations


def kendall_tau(ranking_a: list[str], ranking_b: list[str]) -> float | None:
    """Compute Kendall's Tau-b between two rankings.

    Both rankings must contain the same items. Returns None if fewer than
    2 items are shared between the rankings.
    """
    common = [x for x in ranking_a if x in ranking_b]
    n = len(common)
    if n < 2:
        return None

    rank_b = {item: i for i, item in enumerate(ranking_b)}
    ordered = [rank_b[item] for item in common]

    concordant = 0
    discordant = 0
    for i in range(n):
        for j in range(i + 1, n):
            if ordered[i] < ordered[j]:
                concordant += 1
            elif ordered[i] > ordered[j]:
                discordant += 1

    pairs = n * (n - 1) / 2
    if pairs == 0:
        return None

    return (concordant - discordant) / pairs
